pycaw/magic.py

```diff
@@ -0,0 +1,812 @@
+            raise NotImplementedError(
+                "only one MagicSession wrapper is allowed")
+
+        for iid, magic_root_session in cls.magic_root_sessions.items():
+            magic_session = MagicSessionClass.initialize(
+                magic_root_session, *args, **kwargs)
+            cls.magic_sessions[iid] = magic_session
+        cls.MagicSessionConfigured = (MagicSessionClass, args, kwargs)
+
+    @classmethod
+    def add_magic_app(cls, magic_app, apps_exe):
+        """
+        gets called when a new magic_app is created.
+        this will tell the MagicManager add all current and
+        new matching sessions by app_exe to the magic_app.
+        """
+        if not cls.magic_activated:
+            cls.activate_magic()
+        log.info(f"searching matching active sessions for: {magic_app}")
+        for app_exe in apps_exe:
+            for iid, magic_root_session in cls.magic_root_sessions.items():
+                # and not magic_root_session.magic_app
+                # will prohibit multiple magic_apps to use the same
+                # magic_root_session
+                if (magic_root_session.app_exe == app_exe and
+                        not magic_root_session.magic_app):
+                    log.info(f"{magic_root_session} matched {magic_app}.")
+                    magic_app.add_magic_root_session(iid, magic_root_session)
+
+        # keep reference to magic_app to check later
+        # if new session should be added to this magic_app
+        cls.magic_apps.add(magic_app)
+        log.info(f"{magic_app} added to watchlist. {cls.str()}")
+
+    @classmethod
+    def _match_sess_to_mapp(cls, magic_root_session, iid):
+        log.info(f"searching matching magic_app for: {magic_root_session}")
+        new_app_exe = magic_root_session.app_exe
+        for magic_app in cls.magic_apps:
+            for app_exe in magic_app.app_exe:
+                if app_exe == new_app_exe:
+                    log.info(f"Match {magic_root_session} "
+                             f"{magic_app}")
+                    magic_app.add_magic_root_session(
+                        iid, magic_root_session)
+                    # return will prohibit multiple magic_apps
+                    # to use the same magic_root_session
+                    return
+
+    @classmethod
+    def remove_session(cls, iid, magic_app=None):
+        """magic_root_session will get removed because it is expired"""
+        # pop(iid, None) must not be necessary
+        magic_root_session = cls.magic_root_sessions.pop(iid)
+
+        log.info(f":: removed {magic_root_session}")
+
+        # unregister callback
+        # (magic_root_session -> MagicRootSession -> IAudioSessionEvents)
+        magic_root_session.unregister_notification()
+
+        cls.expired_magic_root_sessions.append(magic_root_session)
+
+        # XXX remove old session:
+        # doesnt work here:
+        # cls.empty_trash()
+
+        """
+        if cls.MagicSessionConfigured:
```

Comment:
commented out code should be removed I guess, isn't it?

Reply:
https://github.com/AndreMiras/pycaw/pull/42#issuecomment-841659891